Spencerkimball/inconsistent reads

spencerkimball opened the pull request:
MVCCGet and MVCCScan now take a "consistent" boolean which, if false,
ignores intents. Moved the guts of MVCCScan into MVCCIterate and MVCCScan
just calls into it. This change allows the removal of the MVCCIterateCommitted
method.

Provides a unittest for "deadloop" issue of encountering an extant intent
at a meta indexing record.

Fixes #343
Fixes #435

spencerkimball committed:
Implementation of alternate "inconsistent" read consistency.

MVCCGet and MVCCScan now take a "consistent" boolean which, if false,
ignores intents. Moved the guts of MVCCScan into MVCCIterate and MVCCScan
just calls into it. This change allows the removal of the MVCCIterateCommitted
method.

Provides a unittest for "deadloop" issue of encountering an extant intent
at a meta indexing record.

Fixes #343
Fixes #435

spencerkimball committed:
Handle interaction of read consistency with multi-range ops.

If read consistency isn't required, a timestamp is assigned using
the local clock. This prevents cases where a different timestamp is
used for each part of a multi-range op, each assigned to the
respective node's clock at the time the RPC is received.

Also, if read consistency isn't required, don't insist that ops
spanning ranges be auto-wrapped in a transaction.

Fixed some errors in dist_sender.go with error handling. In one case
the error wasn't being cleared on the retry loop. In another, the
response error wasn't being examined after calling the RPC.

bdarnell commented on kv/dist_sender_server_test.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+// TestMultiRangeScan verifies that a scan across ranges will
+// return an OpRequiresTxn error.
```
Comment appears to be outdated.

spencerkimball replied:
Fixed.

bdarnell commented on storage/engine/mvcc_test.go:
```diff
@@ -485,17 +485,62 @@ func TestMVCCGetWriteIntentError(t *testing.T) {
 }
 
+// TestMVCCGetInconsistent verifies the behavior of get with
+// inconsistent set to true.
```
s/inconsistent set to true/consistent set to false/

spencerkimball replied:
Done.

bdarnell commented on storage/range.go:
```diff
@@ -273,6 +273,26 @@ func (r *Range) IsLeader() bool {
 	return true
 }
 
+// canServiceCmd returns an error in the event thatthe range replica
```
s/thatthe/that the/

spencerkimball replied:
Done.

bdarnell commented on proto/api.proto:
```diff
@@ -45,6 +45,23 @@ message ClientCmdID {
+  //
+  // TODO(spencer): current unimplemented.
+  CONSENSUS = 1;
+  // INCONSISTENT reads are not guaranteed to read committed data,
```
This sounds like "read uncommitted", but the implementation seems to be returning the latest committed data regardless of any currently pending transactions. Right?

tschottdorf replied:
The comment technically doesn't sound like "read uncommitted", but we could flesh it out a little bit:

```
// INCONSISTENT reads return the latest available, committed value. They are more efficient but may occasionally read stale values since open intents are ignored.
```

spencerkimball replied:
Changed to Tobias' suggested comment.

bdarnell commented:
LGTM

spencerkimball committed:
Allow inconsistent reads to skip the command queue and timestamp cache.

Address pull request feedback and add new unittests for command queue /
timestamp cache skipping functionality.

spencerkimball commented:
Might want to take another look...

spencerkimball committed:
Fixed go generate errors

bdarnell commented on proto/api.proto:
```diff
@@ -45,6 +45,24 @@ message ClientCmdID {
+  // stronger guarantee of consistency than CONSISTENT.
+  //
+  // TODO(spencer): current unimplemented.
+  CONSENSUS = 1;
```
Should we remove the CONSENSUS option until it's implemented? (Do we even need it? When would someone choose one or the other?)

spencerkimball replied:
To pass Jepsen tests which mess with clocks to foil read leases. I'd like to just leave it in as a placeholder.

bdarnell replied:
But is this ever a choice you'd want to make on a per-operation basis? Either you trust your clocks and choose between CONSISTENT and INCONSISTENT, or you don't and choose between CONSENSUS and INCONSISTENT. The choice between CONSISTENT and CONSENSUS seems to be best made at the level of the deployment instead of the operation.

bdarnell commented on storage/engine/mvcc.go:
```diff
@@ -496,14 +499,26 @@ type getValueFunc func(engine Engine, start, end proto.EncodedKey,
+		return nil, util.Errorf("cannot allow inconsistent reads within a transaction")
+	}
 
+	var err error
```
Why create an error variable here?

spencerkimball replied:
Pete's code. Unrelated to my change.

bdarnell commented on storage/range.go:
```diff
@@ -273,6 +274,26 @@ func (r *Range) IsLeader() bool {
+			return util.Errorf("consensus reads not implemented")
+		} else if args.Header().ReadConsistency == proto.INCONSISTENT && args.Header().Txn != nil {
+			return util.Errorf("cannot allow inconsistent reads within a transaction")
+		}
```
This should also return an error if ReadConsistency is not one of the 3 defined constants (which I'm pretty sure the go protobuf implementation allows). Otherwise unknown values will be treated as `!= CONSISTENT`.

bdarnell commented:
LGTM
